ENH: fix numerical issues in roots_jacobi and related special functions that occur for large parameter values

SJamieson opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #14035

#### What does this implement/fix?
Changes computation of weights in roots_jacobi, _gen_roots_and_weights, and roots_gegenbauer to avoid underflow, overflow, and other invalid numerical results. Added unit tests to validate that correct values are now obtained for large values of a, b, and n.

#### Additional information
<!--Any additional information you think is important.-->
This PR complements #12580 -- each PR solves different, but related, problems

SJamieson commented:
Just as a heads up, all of the added tests fail on `master` but pass after these changes. Let me know if I should add comments to the code to explain the changes

person142 commented on scipy/special/orthogonal.py:
```diff
@@ -262,7 +264,10 @@ def roots_jacobi(n, alpha, beta, mu=False):
         return roots_gegenbauer(m, alpha+0.5, mu)
 
-    mu0 = 2.0**(alpha+beta+1)*cephes.beta(alpha+1, beta+1)
+    if (alpha + beta) <= 1000:
```
Certainly a better way to do it, though an even better way would be to use the Lanczos approximation on all the Gamma terms in the Beta function. You'd need to use this private function exposed in `_ufuncs`:

https://github.com/scipy/scipy/blob/master/scipy/special/functions.json#L38

It's ok if you don't want to do that.

SJamieson replied:
Ah, that's a cool idea -- since there's no rush for this to be patched I'm happy to try that. I'll try to push an update within the next few days

SJamieson replied:
@person142 So I went through testing a Lanczos approximation based solution in both places (here and in the other commented place) but I was getting lower accuracy than using the current proposed changes (around a 1e-7 relative error). I think that I might have accidentally neglected some numerical issues with how I was using Lanczos, so a more sophisticated implementation by a more clever developer could probably do better than what we have here, but it does make me a little concerned about the complexity and reduced portability/maintainability of such an implemnetation.

Since the proposed changes are accurate to a relative error of around 1e-13 even for very large values of alpha+beta, I'd like to suggest that we move forward with the current approach -- though I'd be happy to add a comment about the current approach and/or suggesting a switch to use the Lanczos approximation in the future.

person142 replied:
Yes, moving forward with this approach sounds good. It's still very clearly an improvement, which should be the bar for merging.

person142 commented on scipy/special/orthogonal.py:
```diff
@@ -1344,7 +1349,13 @@ def roots_gegenbauer(n, alpha, mu=False):
 
-    mu0 = np.sqrt(np.pi) * cephes.gamma(alpha + 0.5) / cephes.gamma(alpha + 1)
+    if alpha <= 170:
+        mu0 = np.sqrt(np.pi) * cephes.gamma(alpha + 0.5) / cephes.gamma(alpha + 1)
```
And here it would be real easy to use Lanczos; though you could also cast it in terms of `sc.poch`.

SJamieson commented:
A new year, a new push for this PR! @person142 let me know if you have any further thoughts on this or if we can go ahead and proceed. I have rebased it onto main and addressed the new linter requirements.

person142 commented on scipy/special/_orthogonal.py:
```diff
@@ -1484,8 +1490,18 @@ def roots_gegenbauer(n, alpha, mu=False):
+              / _ufuncs.gamma(alpha + 1)
+    else:
+        inv_rt_alpha = 1. / np.sqrt(alpha)
+        mu0 = np.sqrt(np.pi) * (inv_rt_alpha
```
Two things here:

- Let's make sure we're adding commentary on what this asymptotic approximation is
- You want to be using Horner's method to evaluate this polynomial

SJamieson replied:
Great ideas, thanks @person142! The Horner's method version is faster and slightly more accurate. I also added a couple quick comments to the other approximations. New changes are pushed.

person142 reviewed (changes requested):
Some comments on the asymptotic series.

person142 reviewed (approved):
Changes look good! Some CI failures that look unrelated to your changes - maybe try rebasing on master and we'll see what happens?

rgommers reviewed (approved):
CI is happy and this was in good shape already, so in it goes. Thanks @SJamieson and @person142!
